### PR TITLE
fix calling method on generic const fn in impl block

### LINKED ISSUE #79898: Error while calling method on generic const fn in impl block when using `const_trait_impl`
Methods on generic parameters cannot be called inside const fns which are inside an `impl` block.

[playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5672d548cbd29b95657c28dab53b3036)

```rust
#![feature(const_refs_to_cell)]
#![feature(const_fn)]
#![feature(const_precise_live_drops)]
#![feature(const_trait_impl)]
#![allow(incomplete_features)]

trait TestTrait {
    fn test_fn(&self);
}

struct Boo<T> {
    t: T
}

impl<T: TestTrait> Boo<T> {
    const fn foo(t: T) {
        t.test_fn();
        std::mem::forget(t);
    }
}

const fn foo<T: TestTrait>(t: T) {
    t.test_fn();
    std::mem::forget(t);
}

fn main() {}
```

Stderr:

```
error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> src/main.rs:17:9
   |
17 |         t.test_fn();
   |         ^^^^^^^^^^^

error: aborting due to previous error
```

Related to: #67792

### PR DESCRIPTION
**Pull request by vn-ki**
fixes #79898

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 1]
```diff
@@ -789,6 +789,7 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
 
                     // If the method is provided via a where-clause that does not use the `?const`
                     // opt-out, the call is allowed.
+                    debug!("cconst_trait_impl: implsc={:?}", implsrc);
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:792]:
```suggestion
                    debug!("const_trait_impl: implsc={:?}", implsrc);
```

**Review by oli-obk — COMMENTED** [review]
Awe.Some! I love how simple it ended up being. Two things that come to mind

1. we should have a test for it, maybe just extend one of the existing ones for bare `const fn` with method impls?
2. Several `GenericPredicates` constructor expressions use `parent: None, constness: NotConst`. Not sure if that is better, but it could be changed to `.. Default::default()` syntax. What do you think? I have no strong opinion about it, just noticed that it was possible.

**Comment by crlf0710** [comment]
@vn-ki Ping from triage! CI is still red here. Would you mind fixing that?

**Comment by vn-ki** [comment]
This method does not work. Closing.

**Pull request closed without merging**